src/v/ssx/when_all.h

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include <seastar/core/future.hh>
+#include <seastar/coroutine/as_future.hh>
+
+#include <concepts>
+#include <ranges>
+#include <type_traits>
+
+namespace ssx {
+
+namespace detail {
+
+template<typename Container>
+concept emplace_backable = requires(Container c) {
+    typename Container::value_type;
+    c.emplace_back(std::declval<typename Container::value_type>());
+};
```

Comment:
`push_back` vs `emplace_back`: yes, it felt like an easy way to get a little bit more flexibility on what is allowed, conversions included.

on the `emplace_backable`:
This is also touched on a couple more comments, so i am replying here for everything.
The things that is actually required is to be able to add an element out of whatever comes out of the futures.

Looking at it again, this `emplace_backable` is asking if you can copy-emplace.

The combination of both of these tests is to ensure the compatibility of the two containers.
Instead of this `emplace_backable` and `convertible` it seems like a better idea to explicitly ask if the future::value_type is emplace-able to the output container, using your proposed version of `emplace_backable`. I will re-work it.

Reply:
My understanding about `emplace_backable` was somewhat flawed. I was expecting it that it would verify "in-depth" if it was valid. It turns out that
```
emplace_backable<std::vector<std::string>, int>
```
only checks that the expression
```
std::vector<std::string> vec;
vec.template emplace_back<int>(...);
```
is valid. Which is. It's the instantiation of this function that fails.

To protect against the above scenario, i kept the `std::constructible_from` check.

Let me know what you think about it.